<!-- 
Copyright (c) 2019  Carnegie Mellon University
 
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
 
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
 
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE. 
-->

<launch>
  <arg name="use_keyboard" default="false"/>
  <arg name="use_joy" default="true"/>
  <arg name="use_imu" default="true"/>
  
  <arg name="max_acc" default="1.0"/>
  <arg name="target_rate" default="20"/>
  <arg name="bias" default="0.258"/>
  <arg name="gain_omega" default="2.0"/>
  <arg name="gain_omega_i" default="0.0"/>

  <arg name="wheel_diameter" default="0.073"/>
  <arg name="count_per_round" default="8192"/>
  <arg name="left_is_1" default="false"/>
  <arg name="gain_left" default="1.0"/>
  <arg name="gain_right" default="1.0"/>
  
  <arg name="vel_gain" default="2.25" />
  <arg name="vel_integrator_gain" default="22.5" />
  <arg name="encoder_bandwidth" default="200" />
  <arg name="motor_bandwidth" default="200" />
  
  <arg name="joy_dev" default="/dev/input/js0" />
  <arg name="config_filepath" default="$(find motor_adapter)/config/gamepad.config.yaml" />

  
  <node pkg="nodelet" type="nodelet" output="screen" name="cabot_nodelet_manager"
	args="manager"/>
  
  <node pkg="nodelet" type="nodelet" output="screen" name="odriver_adapter_node"
	args="load MotorAdapter/ODriverNodelet cabot_nodelet_manager">
    <param name="encoder_topic" type="string" value="/motorStatus" />
    <param name="odom_topic" type="string" value="/odom" />
    
    <param name="motor_topic" type="string" value="/motorTarget" />
    <param name="cmd_vel_topic" type="string" value="/cmd_vel" />
    
    <param name="max_acc" value="$(arg max_acc)"/>
    <param name="target_rate" value="$(arg target_rate)"/>
    <param name="bias" value="$(arg bias)"/>

    <param name="gain_omega" value="$(arg gain_omega)" if="$(arg use_imu)" />
    <param name="gain_omega_i" value="$(arg gain_omega_i)" if="$(arg use_imu)" />
  </node>
	  
  <node pkg="odriver" type="odriver_node.py" name="odriver_node" output="screen">
    <param name="port" value="/dev/ttyODRIVE" />

    <param name="wheel_diameter" value="$(arg wheel_diameter)" />
    <param name="count_per_round" value="$(arg count_per_round)" />
    <param name="left_is_1" value="$(arg left_is_1)" />

     <!-- adjust motor speed -->
    <param name="gain_left" value="$(arg gain_left)" />
    <param name="gain_right" value="$(arg gain_right)" />

    <!-- adjust motor control config -->
    <param name="vel_gain" value="$(arg vel_gain)" />
    <param name="vel_integrator_gain" value="$(arg vel_integrator_gain)" />
    <param name="encoder_bandwidth" value="$(arg encoder_bandwidth)" />
    <param name="motor_bandwidth" value="$(arg motor_bandwidth)" />
    
    <remap from="/motorTarget" to="/motorTarget" />
    <remap from="/motorStatus" to="/motorStatus" />
  </node>

  <group if="$(arg use_joy)">
    <node pkg="joy" type="joy_node" name="joy_node">
      <param name="dev" value="$(arg joy_dev)" />
      <param name="deadzone" value="0.3" />
      <param name="autorepeat_rate" value="20" />
    </node>
    <node pkg="teleop_twist_joy" name="teleop_twist_joy" type="teleop_node">
      <rosparam command="load" file="$(arg config_filepath)" />
    </node>
  </group>

  <group if="$(arg use_keyboard)">
    <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py"
	  name="teleop_keyboard" output="screen">
    </node>
  </group>

  <group if="$(arg use_imu)">
    <node pkg="cabot" type="cabot_serial.py" name="rosserial" output="log">
      <param name="port" value="/dev/ttyARDUINO_MEGA"/>
      <param name="baud" value="115200"/>
    </node>
  </group>
</launch>
